ValidatorCleanRestart: Re-enable the first test

hewison-chris opened the pull request:
This test had been disabled but could be hiding a problem.

hewison-chris committed:
ValidatorCleanRestart: Re-enable the first test

The test also required some changes as it had been disabled for some
time and was out of sync with the implementation code. The actual change
in the test that prevents it from being unreliable is waiting for the
new set of validators to complete network discovery before generating
the new block.

linked0 commented on source/agora/test/ValidatorCleanRestart.d:
```diff
@@ -31,10 +31,10 @@ import agora.test.Base;
 version(none) unittest
```
Shouldn't this line be changed in order to re-enable the first test?

hewison-chris replied:
Yes, thanks

mkykadir commented on source/agora/test/Base.d:
```diff
@@ -1331,6 +1331,10 @@ public class TestNetworkManager : NetworkManager
         Duration timeout)
     {
         auto tid = this.registry.locate!TestAPI(address.host);
+        if (tid != typeof(tid).init)
+            return new RemoteAPI!TestAPI(tid, timeout);
+        // Try again
+        tid = this.registry.locate!TestAPI(address.host);
```
Separated commit, maybe

hewison-chris replied:
I'll remove as I think there is another solution pending in a different PR.

mkykadir commented on source/agora/test/ValidatorCleanRestart.d:
```diff
@@ -93,16 +89,16 @@ version(none) unittest
     // Check if the validators in the set B have all the addresses for
     // current validators and previous validators except themselves.
```
```suggestion
    // current validators and previous validators.
```
since we're checking `>=` now

hewison-chris replied:
I will update the comment and the check to be an exact count once I double check what it should be.

hewison-chris replied:
```
["http://validator-0.boa1xrval5rzmma29zh4aqgv3mvcarhwa0w8rgthy3l9vaj3fywf9894ycmjkm8.localrest/", 
"http://validator-3.boa1xrval6hd8szdektyz69fnqjwqfejhu4rvrpwlahh9rhaazzpvs5g6lh34l5.localrest/", 
"http://validator-2.boa1xrval7gwhjz4k9raqukcnv2n4rl4fxt74m2y9eay6l5mqdf4gntnzhhscrh.localrest/", 
"http://validator-1.boa1xzval3ah8z7ewhuzx6mywveyr79f24w49rdypwgurhjkr8z2ke2mycftv9n.localrest/", 
"http://validator-6.boa1xza007gllhzdawnr727hds36guc0frnjsqscgf4k08zqesapcg3uujh9g93.localrest/", 
"http://validator-5.boa1xzval2a3cdxv28n6slr62wlczslk3juvk7cu05qt3z55ty2rlfqfc6egsh2.localrest/", 
"http://validator-8.boa1xqd00qsu7n5ykyckc23wmcjglfalcdea3x2af88hx2x5qx65x7w8g2r5t29.localrest/", 
"http://name.registry/", 
"http://validator-4.boa1xzval4nvru2ej9m0rptq7hatukkavemryvct4f8smyy3ky9ct5u0s8w6gfy.localrest/", 
"http://validator-7.boa1xrc00kar2yqa3jzve9cm4cvuaa8duazkuwrygmqgpcuf0gqww8ye7ua9lkl.localrest/"]
```
It should include the registry and all validators.

omerfirmak commented on source/agora/consensus/Fee.d:
```diff
@@ -786,7 +786,7 @@ unittest
 
     Transaction freeze_tx = Transaction(
-        [ Output(Amount(2_000_000L * 10_000_000L), WK.Keys.NODE2.address, OutputType.Freeze) ]);
+        [ Output(Amount(2_000_000.coins), WK.Keys.NODE2.address, OutputType.Freeze) ]);
```
just `2_000_000.coins` would work fine, they are of type `Amount` anyways

hewison-chris replied:
Well spotted.

mkykadir commented on source/agora/test/ValidatorCleanRestart.d:
```diff
@@ -93,16 +89,16 @@ version(none) unittest
-    network.generateBlocks(iota(GenesisValidators, cast(size_t) nodes.length),
-        Height(GenesisValidatorCycle + 1));
+    network.generateBlocks(iota(GenesisValidators, all_validators),
+        Height(GenesisValidatorCycle + 1), 20.seconds);
```
That's why CI is not compiling, duration

hewison-chris replied:
Thanks, wrong function for that param.

hewison-chris commented:
I think this is now actually failing because we need to fix something.

hewison-chris commented:
Now updated to `waitForDiscovery` of the `outsider validators` after they are restarted before creating the block.

omerfirmak commented on source/agora/test/ValidatorCleanRestart.d:
```diff
@@ -48,24 +50,18 @@ version(none) unittest
-    auto to_send = txb.draw(Amount.MinFreezeAmount, keys).sign(OutputType.Freeze);
-    set_a.each!(n => n.postTransaction(to_send));
+    network.postAndEnsureTxInPool(iota(GenesisValidators),
+        network.freezeUTXO(iota(GenesisValidators, GenesisValidators + 3)));
```
```suggestion
        network.freezeUTXO(iota(GenesisValidators, all_validators)));
```

hewison-chris replied:
Updated

omerfirmak commented on source/agora/test/ValidatorCleanRestart.d:
```diff
@@ -91,17 +87,22 @@ version(none) unittest
-    set_a.each!(node => node.ctrl.sleep(6.seconds, true));
+    set_a.each!(node => node.ctrl.shutdown);
+
+    network.waitForDiscovery(only(6, 7, 8));
```
```suggestion
    network.waitForDiscovery(iota(GenesisValidators, all_validators));
```

isnt this the same?

hewison-chris replied:
Updated

hewison-chris commented:
~Waiting for https://github.com/bosagora/agora/pull/2859~
```
2022-01-17 07:01:53,569 Info [agora.network.Manager] - Couldn't register our address: geod24.LocalRest.ClientException@submodules/localrest/source/geod24/LocalRest.d-mixin-1217(1236): Request to void geod24.LocalRest.RemoteAPI!(NameRegistryAPI, Serializer).RemoteAPI.postValidator(RegistryPayload _param_0) @safe couldn't be processed : agora.common.Ensure.FormattedException@source/agora/node/Registry.d(245): Not an enrolled validator```

linked0 reviewed (approved):
LGTM

hewison-chris committed:
Base Test class: Add some named constants

This is just to make it clearer what we are counting.

hewison-chris committed:
Add some debug logging

Some of the debug log lines are not `nothrow` so for convenience the
formatting of the debug log message is wrapped in an `assumeWontThrow`
which will `assert` if an `exception` is `thrown` and report it in the
assertion message.

hewison-chris committed:
Unit test: Reduce nodes in ValidatorCleanRestart

By reducing the number of nodes to 4 it is easier to analyze the logs
when the test fails. It should also help it be more reliable in ci.

hewison-chris committed:
Update identity of self when utxo is known

After a clean restart of a validator which has a stake frozen after the
GenesisBlock. In this case when blocks are accepted it can check for a
frozen utxo and update the identity when found.
